From e9f81b85281d578c83d50e271c33625f2b8fa51f Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 8 Jul 2024 20:51:30 +0200
Subject: [PATCH] Add legacy mode for Script execution without BCH upgrades
 (#19)

Dogecoin doesn't have a lot of the script features that BCH/XEC added. Instead of removing them (and having to refactor a lot of code), we add a script flag SCRIPT_VERIFY_LEGACY_RULES, which runs the interpreter in legacy mode:

- Disable OP_CAT, OP_SPLIT, OP_AND, OP_OR, OP_XOR, OP_DIV, OP_MOD, OP_NUM2BIN, OP_BIN2NUM.
- Remove OP_REVERSEBYTES.
- Remove OP_CHECKDATASIG and OP_CHECKDATASIGVERIFY.
- Disable Schnorr signatures.

diff --git a/src/consensus/params.h b/src/consensus/params.h
index 5321f85398e31ed6888df26bc8c9c86f074187bc..1b7f99b4868b7caf5a6f4d9e71c208a9fc702711 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -98,6 +98,8 @@ struct Params {
 
     /** Enforce Dogecoin chain ID for AuxPoW headers. */
     bool enforceStrictAuxPowChainId;
+    /** Whether to enforce the old legacy Script rules. */
+    bool enforceLegacyScriptRules;
 
     /** Enable or disable the miner fund by default */
     bool enableMinerFund;
diff --git a/src/init.cpp b/src/init.cpp
index f817aed2fe09e71f1d05e9c695b65fa77cd939fb..54a4598a541a761ee039f396e0e68533469f72bf 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -500,6 +500,7 @@ void SetupServerArgs(NodeContext &node) {
         "-parkdeepreorg",
         "-automaticunparking",
         "-replayprotectionactivationtime",
+        "-legacyscriptrules",
         "-enableminerfund",
         "-chronikallowpause",
         "-chronikcors",
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index 94b25a2466bcdcfdf781edf0649e4e08f7d8bf7d..0e38bee43a9a7ffe3c98ce709c3fa77336b13ae9 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -137,6 +137,8 @@ public:
         consensus.digishieldHeight = 145000;
         consensus.digishieldMinDiffHeight = 157500;
         consensus.initialCoinbaseMaturity = 30;
+        // Dogecoin: Enforce legacy script rules on mainnet
+        consensus.enforceLegacyScriptRules = true;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -295,6 +297,8 @@ public:
         consensus.digishieldHeight = 145000;
         consensus.digishieldMinDiffHeight = 157500;
         consensus.initialCoinbaseMaturity = 30;
+        // Dogecoin: Enforce legacy script rules on testnet
+        consensus.enforceLegacyScriptRules = true;
 
         diskMagic[0] = 0xfb;
         diskMagic[1] = 0x87;
@@ -453,6 +457,8 @@ public:
         consensus.digishieldMinDiffHeight = 157500;
         // keep maturity same as Bitcoin for tests
         consensus.initialCoinbaseMaturity = REGTEST_COINBASE_MATURITY;
+        // legacy rules disabled for regtest so we don't refactor the universe
+        consensus.enforceLegacyScriptRules = false;
 
         diskMagic[0] = 0x94;
         diskMagic[1] = 0xb1;
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index c9c01590f50c7a47b2b07ecc8c3f28de47520c9a..9721c056a1fe18f8193182b1d9bb4130e17fd0d2 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -91,6 +91,17 @@ static bool IsOpcodeDisabled(opcodetype opcode, uint32_t flags) {
             // Disabled opcodes.
             return true;
 
+        case OP_CAT:
+        case OP_SPLIT:
+        case OP_NUM2BIN:
+        case OP_BIN2NUM:
+        case OP_AND:
+        case OP_OR:
+        case OP_XOR:
+        case OP_DIV:
+        case OP_MOD:
+            return (flags & SCRIPT_VERIFY_LEGACY_RULES) != 0;
+
         default:
             break;
     }
@@ -965,6 +976,9 @@ bool ScriptInterpreter::RunNextOpInner() {
 
             case OP_CHECKDATASIG:
             case OP_CHECKDATASIGVERIFY: {
+                if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+                    return set_error(serror, ScriptError::BAD_OPCODE);
+                }
                 // (sig message pubkey -- bool)
                 if (stack.size() < 3) {
                     return set_error(serror,
@@ -988,7 +1002,7 @@ bool ScriptInterpreter::RunNextOpInner() {
                         .Write(vchMessage.data(), vchMessage.size())
                         .Finalize(vchHash.data());
                     fSuccess = checker.VerifySignature(
-                        vchSig, CPubKey(vchPubKey), uint256(vchHash));
+                        vchSig, CPubKey(vchPubKey), uint256(vchHash), flags);
                     metrics.nSigChecks += 1;
 
                     if (!fSuccess && (flags & SCRIPT_VERIFY_NULLFAIL)) {
@@ -1280,6 +1294,9 @@ bool ScriptInterpreter::RunNextOpInner() {
             } break;
 
             case OP_REVERSEBYTES: {
+                if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+                    return set_error(serror, ScriptError::BAD_OPCODE);
+                }
                 // (in -- out)
                 if (stack.size() < 1) {
                     return set_error(serror,
@@ -1653,8 +1670,8 @@ uint256 SignatureHash(const CScript &scriptCode, const T &txTo,
 
 bool BaseSignatureChecker::VerifySignature(const std::vector<uint8_t> &vchSig,
                                            const CPubKey &pubkey,
-                                           const uint256 &sighash) const {
-    if (vchSig.size() == 64) {
+                                           const uint256 &sighash, uint32_t flags) const {
+    if (vchSig.size() == 64 && (flags & SCRIPT_VERIFY_LEGACY_RULES) == 0) {
         return pubkey.VerifySchnorr(sighash, vchSig);
     } else {
         return pubkey.VerifyECDSA(sighash, vchSig);
@@ -1681,7 +1698,7 @@ bool GenericTransactionSignatureChecker<T>::CheckSig(
     uint256 sighash = SignatureHash(scriptCode, *txTo, nIn, sigHashType, amount,
                                     this->txdata, flags, &m_sighash_cache);
 
-    if (!VerifySignature(vchSig, pubkey, sighash)) {
+    if (!VerifySignature(vchSig, pubkey, sighash, flags)) {
         return false;
     }
 
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index 34c35b9f58e6c11ac35384ceee627ec6aebc9e4e..4457d995e2ea93cf91dda3c276b5e64fd61626ea 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -75,7 +75,7 @@ class BaseSignatureChecker {
 public:
     virtual bool VerifySignature(const std::vector<uint8_t> &vchSig,
                                  const CPubKey &vchPubKey,
-                                 const uint256 &sighash) const;
+                                 const uint256 &sighash, uint32_t flags) const;
 
     virtual bool CheckSig(const std::vector<uint8_t> &vchSigIn,
                           const std::vector<uint8_t> &vchPubKey,
diff --git a/src/script/script_flags.h b/src/script/script_flags.h
index 7571d2f397c928552bf9e9c322429109def460e8..7304f33c700b0c4d7d5497e1d4d30780a64b3cc2 100644
--- a/src/script/script_flags.h
+++ b/src/script/script_flags.h
@@ -104,6 +104,14 @@ enum {
 
     // A utility flag to decide whether we must enforce sigcheck limits.
     SCRIPT_ENFORCE_SIGCHECKS = (1U << 23),
+
+    // Whether to run the interpreter in legacy mode (i.e. like on Dogecoin)
+    // - Disable OP_CAT, OP_SPLIT, OP_AND, OP_OR, OP_XOR, OP_DIV, OP_MOD,
+    //   OP_NUM2BIN, OP_BIN2NUM.
+    // - Remove OP_REVERSEBYTES.
+    // - Remove OP_CHECKDATASIG and OP_CHECKDATASIGVERIFY.
+    // - Disable Schnorr signatures.
+    SCRIPT_VERIFY_LEGACY_RULES = (1U << 24),
 };
 
 #endif // BITCOIN_SCRIPT_SCRIPT_FLAGS_H
diff --git a/src/script/sigcache.cpp b/src/script/sigcache.cpp
index f2c834e8ff04dcaa7329f1e7741f1ceb79859a9c..c2a21addf6e4a48ebbc73c3c1bbe6c6c29744b03 100644
--- a/src/script/sigcache.cpp
+++ b/src/script/sigcache.cpp
@@ -79,10 +79,10 @@ bool CachingTransactionSignatureChecker::IsCached(
 
 bool CachingTransactionSignatureChecker::VerifySignature(
     const std::vector<uint8_t> &vchSig, const CPubKey &pubkey,
-    const uint256 &sighash) const {
+    const uint256 &sighash, uint32_t flags) const {
     return RunMemoizedCheck(
         m_signature_cache, vchSig, pubkey, sighash, store, [&] {
             return TransactionSignatureChecker::VerifySignature(vchSig, pubkey,
-                                                                sighash);
+                                                                sighash, flags);
         });
 }
diff --git a/src/script/sigcache.h b/src/script/sigcache.h
index b9c038e90e5ecc561386553c1708dc06ab9c5470..2b3af669cf1012339aafb44d46d051e00558e63f 100644
--- a/src/script/sigcache.h
+++ b/src/script/sigcache.h
@@ -73,8 +73,8 @@ public:
           store(storeIn), m_signature_cache(signature_cache) {}
 
     bool VerifySignature(const std::vector<uint8_t> &vchSig,
-                         const CPubKey &vchPubKey,
-                         const uint256 &sighash) const override;
+                         const CPubKey &vchPubKey, const uint256 &sighash,
+                         uint32_t flags) const override;
 
     friend class TestCachingTransactionSignatureChecker;
 };
diff --git a/src/test/checkdatasig_tests.cpp b/src/test/checkdatasig_tests.cpp
index 4a984c540f8dd099b2a321a622c853a80825d3c3..1bd9373868a206b8910ff1ddeb93bf4550551bdc 100644
--- a/src/test/checkdatasig_tests.cpp
+++ b/src/test/checkdatasig_tests.cpp
@@ -141,6 +141,15 @@ BOOST_AUTO_TEST_CASE(checkdatasig_test) {
     for (int i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
 
+        if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+            // Legacy has OP_CHECKDATASIG and OP_CHECKDATASIGVERIFY disabled
+            CheckError(flags, {nondersig, message, pubkeyC}, script,
+                       ScriptError::BAD_OPCODE);
+            CheckError(flags, {nondersig, message, pubkeyC}, scriptverify,
+                       ScriptError::BAD_OPCODE);
+            continue;
+        }
+
         if (flags & SCRIPT_VERIFY_STRICTENC) {
             // When strict encoding is enforced, hybrid keys are invalid.
             CheckError(flags, {{}, message, pubkeyH}, script,
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 9af09a3e2276fc61e4e9ad469426173c83bc894b..e511cbad5b03906805fce22b8a6503c3a54685ce 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -3936,5 +3936,33 @@
 ["0 0x09 0x300602010102010141", "1 0x21 0x02865c40293a680cb9c020e7b1e106d8c1916d3cef99aa431a56d253e69256dac0 1 CHECKMULTISIG NOT", "STRICTENC", "ILLEGAL_FORKID"],
 ["0 0x09 0x300602010102010141", "1 0x21 0x02865c40293a680cb9c020e7b1e106d8c1916d3cef99aa431a56d253e69256dac0 1 CHECKMULTISIG NOT", "SIGHASH_FORKID", "OK"],
 
+["LEGACY SCRIPT RULES"],
+["1 2", "OP_CAT", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_SPLIT", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_AND", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_OR", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_XOR", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_DIV", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_MOD", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_NUM2BIN", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_BIN2NUM", "LEGACY_RULES", "DISABLED_OPCODE"],
+["1 2", "OP_REVERSEBYTES", "LEGACY_RULES", "BAD_OPCODE"],
+["1 2", "OP_CHECKDATASIG", "LEGACY_RULES", "BAD_OPCODE"],
+["1 2", "OP_CHECKDATASIGVERIFY", "LEGACY_RULES", "BAD_OPCODE"],
+[
+    "0x41 0x0df4be7f5fe74b2855b92082720e889038e15d8d747334fa3f300ef4ab1db1eea56aa83d1d60809ff6703791736be87cfb6cbc5c4036aeed3b4ea4e6dab3509001",
+    "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
+    "LEGACY_RULES,STRICTENC",
+    "EVAL_FALSE",
+    "CHECKSIG Schnorr w/ STRICTENC disallowed on legacy"
+],
+[
+    "0x41 0x8d37e95e36718d7fbc8ffd63b0b4ebb89dd5fb683510a95345869399f810a8724a00e5cbbb2190205ffabac601d80bf110a67013521b7a7c02b2a51e07d723eb01",
+    "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIGVERIFY 1",
+    "LEGACY_RULES,STRICTENC",
+    "CHECKSIGVERIFY",
+    "CHECKSIGVERIFY Schnorr w/ STRICTENC disallowed on legacy"
+],
+
 ["The End"]
 ]
diff --git a/src/test/fuzz/script_sigcache.cpp b/src/test/fuzz/script_sigcache.cpp
index 896eda41d246f558224587779c2b436d436e2232..542d7bebaf98009405dcbf8bf25dc35f0fecb76b 100644
--- a/src/test/fuzz/script_sigcache.cpp
+++ b/src/test/fuzz/script_sigcache.cpp
@@ -56,7 +56,8 @@ FUZZ_TARGET_INIT(script_sigcache, initialize_script_sigcache) {
             ConsumeRandomLengthByteVector(fuzzed_data_provider);
         if (!random_bytes.empty()) {
             (void)caching_transaction_signature_checker.VerifySignature(
-                random_bytes, *pub_key, ConsumeUInt256(fuzzed_data_provider));
+                random_bytes, *pub_key, ConsumeUInt256(fuzzed_data_provider),
+                0);
         }
     }
 }
diff --git a/src/test/op_reversebytes_tests.cpp b/src/test/op_reversebytes_tests.cpp
index 7dfbe66d8a8260dbaa82482fb1391f7ac53a7691..2f146d3739482e94673db38cde0bdf9178f4989b 100644
--- a/src/test/op_reversebytes_tests.cpp
+++ b/src/test/op_reversebytes_tests.cpp
@@ -68,6 +68,11 @@ BOOST_AUTO_TEST_CASE(op_reversebytes_manual_random_flags) {
     MMIXLinearCongruentialGenerator lcg;
     for (size_t i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
+        if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+            // Legacy has OP_REVERSEBYTES disabled
+            continue;
+        }
+
         CheckPassForCombinations(flags, {}, {});
         CheckPassForCombinations(flags, {99}, {99});
         CheckPassForCombinations(flags, {0xde, 0xad}, {0xad, 0xde});
@@ -91,6 +96,11 @@ BOOST_AUTO_TEST_CASE(op_reversebytes_iota) {
         valtype iota_data_reversed = {iota_data.rbegin(), iota_data.rend()};
         for (size_t i = 0; i < 4096; i++) {
             uint32_t flags = lcg.next();
+            if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+                // Legacy has OP_REVERSEBYTES disabled
+                continue;
+            }
+
             CheckPassForCombinations(flags, iota_data, iota_data_reversed);
         }
     }
@@ -131,6 +141,11 @@ BOOST_AUTO_TEST_CASE(op_reversebytes_random_and_palindrome) {
         }
 
         for (const uint32_t flags : flaglist) {
+            if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+                // Legacy has OP_REVERSEBYTES disabled
+                continue;
+            }
+
             // Verify random data passes.
             CheckPassForCombinations(flags, random_data, random_data_reversed);
             // Verify palindrome check passes.
@@ -146,6 +161,13 @@ BOOST_AUTO_TEST_CASE(op_reversebytes_failures) {
     for (size_t i = 0; i < 4096; i++) {
         uint32_t flags = lcg.next();
 
+        if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+            // Legacy has OP_REVERSEBYTES disabled
+            CheckErrorWithFlags(flags, {}, CScript() << OP_REVERSEBYTES,
+                                ScriptError::BAD_OPCODE);
+            continue;
+        }
+
         // Verify non-palindrome fails.
         CheckErrorWithFlags(flags, {{0x01, 0x02, 0x03, 0x02, 0x02}},
                             CScript()
diff --git a/src/test/schnorr_tests.cpp b/src/test/schnorr_tests.cpp
index 229f0a08367f4dbc65d4b228e78664375dbc0b92..fc42fe742f57f1936df87834c8a0675fae60fb5b 100644
--- a/src/test/schnorr_tests.cpp
+++ b/src/test/schnorr_tests.cpp
@@ -106,6 +106,7 @@ BOOST_AUTO_TEST_CASE(opcodes_random_flags) {
 
         const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
         const bool hasNullFail = (flags & SCRIPT_VERIFY_NULLFAIL) != 0;
+        const bool isLegacy = (flags & SCRIPT_VERIFY_LEGACY_RULES) != 0;
 
         // Prepare 65-byte transaction sigs with right hashtype byte.
         valtype DER64_with_hashtype =
@@ -121,17 +122,25 @@ BOOST_AUTO_TEST_CASE(opcodes_random_flags) {
             CheckError(flags, {Zero64_with_hashtype, pubkeyC},
                        scriptCHECKSIGVERIFY, ScriptError::SIG_NULLFAIL);
             CheckError(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIG,
-                       ScriptError::SIG_NULLFAIL);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::SIG_NULLFAIL);
             CheckError(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIGVERIFY,
-                       ScriptError::SIG_NULLFAIL);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::SIG_NULLFAIL);
         } else {
             CheckPass(flags, {Zero64_with_hashtype, pubkeyC}, scriptCHECKSIG,
                       {});
             CheckError(flags, {Zero64_with_hashtype, pubkeyC},
                        scriptCHECKSIGVERIFY, ScriptError::CHECKSIGVERIFY);
-            CheckPass(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIG, {});
+            if (isLegacy) {
+                CheckError(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIG,
+                           ScriptError::BAD_OPCODE);
+            } else {
+                CheckPass(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIG, {});
+            }
             CheckError(flags, {Zero64, {}, pubkeyC}, scriptCHECKDATASIGVERIFY,
-                       ScriptError::CHECKDATASIGVERIFY);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::CHECKDATASIGVERIFY);
         }
 
         // Test CHECKSIG & CHECKDATASIG with DER sig, which fails upon
@@ -142,17 +151,25 @@ BOOST_AUTO_TEST_CASE(opcodes_random_flags) {
             CheckError(flags, {DER64_with_hashtype, pubkeyC},
                        scriptCHECKSIGVERIFY, ScriptError::SIG_NULLFAIL);
             CheckError(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIG,
-                       ScriptError::SIG_NULLFAIL);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::SIG_NULLFAIL);
             CheckError(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIGVERIFY,
-                       ScriptError::SIG_NULLFAIL);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::SIG_NULLFAIL);
         } else {
             CheckPass(flags, {DER64_with_hashtype, pubkeyC}, scriptCHECKSIG,
                       {});
             CheckError(flags, {DER64_with_hashtype, pubkeyC},
                        scriptCHECKSIGVERIFY, ScriptError::CHECKSIGVERIFY);
-            CheckPass(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIG, {});
+            if (isLegacy) {
+                CheckError(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIG,
+                           ScriptError::BAD_OPCODE);
+            } else {
+                CheckPass(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIG, {});
+            }
             CheckError(flags, {DER64, {}, pubkeyC}, scriptCHECKDATASIGVERIFY,
-                       ScriptError::CHECKDATASIGVERIFY);
+                       isLegacy ? ScriptError::BAD_OPCODE
+                                : ScriptError::CHECKDATASIGVERIFY);
         }
 
         // test OP_CHECKMULTISIG/VERIFY
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index f2049474ff3c0d1bbf0cf7a3d5ed14852b4d3bbe..8c1e57aeee67c9ef258965ec063d04868b230333 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -154,7 +154,7 @@ struct ScriptTest : BasicTestingSetup {
             // as-is.
             extra_flags &= ~(SCRIPT_ENABLE_SIGHASH_FORKID |
                              SCRIPT_ENABLE_REPLAY_PROTECTION |
-                             SCRIPT_ENABLE_SCHNORR_MULTISIG);
+                             SCRIPT_ENABLE_SCHNORR_MULTISIG | SCRIPT_VERIFY_LEGACY_RULES);
             uint32_t combined_flags =
                 expect ? (flags & ~extra_flags) : (flags | extra_flags);
             // Weed out invalid flag combinations.
diff --git a/src/test/scriptflags.cpp b/src/test/scriptflags.cpp
index cc6ad88bf55dfe9c1776d5e26d8c0afecb148f1f..e9a5b4a396a704f951fc41c0befbe17bc99fc273 100644
--- a/src/test/scriptflags.cpp
+++ b/src/test/scriptflags.cpp
@@ -33,6 +33,7 @@ static std::map<std::string, uint32_t> mapFlagNames = {
     {"DISALLOW_SEGWIT_RECOVERY", SCRIPT_DISALLOW_SEGWIT_RECOVERY},
     {"SCHNORR_MULTISIG", SCRIPT_ENABLE_SCHNORR_MULTISIG},
     {"INPUT_SIGCHECKS", SCRIPT_VERIFY_INPUT_SIGCHECKS},
+    {"LEGACY_RULES", SCRIPT_VERIFY_LEGACY_RULES},
 };
 
 uint32_t ParseScriptFlags(std::string strFlags) {
diff --git a/src/test/sigcache_tests.cpp b/src/test/sigcache_tests.cpp
index af950ac85febbaef47789026e3813c130a89f776..c3013757d198a4d0d15c21ce02d527b0c1fdefc6 100644
--- a/src/test/sigcache_tests.cpp
+++ b/src/test/sigcache_tests.cpp
@@ -39,7 +39,7 @@ public:
 
     inline bool VerifyAndStore(const std::vector<uint8_t> &vchSig,
                                const CPubKey &pubkey, const uint256 &sighash) {
-        return pchecker->VerifySignature(vchSig, pubkey, sighash);
+        return pchecker->VerifySignature(vchSig, pubkey, sighash, 0);
     }
 
     inline bool IsCached(const std::vector<uint8_t> &vchSig,
diff --git a/src/test/sigcheckcount_tests.cpp b/src/test/sigcheckcount_tests.cpp
index 6a3e4e940e37df8a3be59b26367e03eadf2f9d1f..8c81669c5c93aa51e0a3ba58be3a327164691e42 100644
--- a/src/test/sigcheckcount_tests.cpp
+++ b/src/test/sigcheckcount_tests.cpp
@@ -57,8 +57,8 @@ static class : public BaseSignatureChecker {
      * value verify false as well. Otherwise, checks verify as true.
      */
     bool VerifySignature(const std::vector<uint8_t> &vchSig,
-                         const CPubKey &vchPubKey,
-                         const uint256 &sighash) const final {
+                         const CPubKey &vchPubKey, const uint256 &sighash,
+                         uint32_t flags) const final {
         if (vchPubKey == CPubKey(badpub)) {
             return false;
         }
diff --git a/src/validation.cpp b/src/validation.cpp
index 3fd8cec2b40a70d3a9c8db8e40ca8e6de0b8ac6e..22dfebb65bf98da284a0b1cda57ad02ef5e59ea0 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -235,6 +235,13 @@ IsReplayProtectionEnabled(const Consensus::Params &params,
            activation_time.value_or(params.sandersActivationTime);
 }
 
+// Command-line argument "-legacyscriptrules" will make the node enforce the old
+// script rules (see SCRIPT_VERIFY_LEGACY_RULES).
+static bool IsLegacyScriptRulesEnabled(const Consensus::Params &params) {
+    return gArgs.GetBoolArg("-legacyscriptrules",
+                            params.enforceLegacyScriptRules);
+}
+
 /**
  * Checks to avoid mempool polluting consensus critical paths since cached
  * signature and script validity results will be reused if we validate this
@@ -2135,6 +2142,11 @@ static uint32_t GetNextBlockScriptFlags(const CBlockIndex *pindex,
         flags |= SCRIPT_ENABLE_REPLAY_PROTECTION;
     }
 
+    // If we are on a legacy network (i.e. Dogecoin), keep the old Script rules
+    if (IsLegacyScriptRulesEnabled(consensusparams)) {
+        flags |= SCRIPT_VERIFY_LEGACY_RULES;
+    }
+
     return flags;
 }
 
diff --git a/test/functional/feature_legacy_script_rules.py b/test/functional/feature_legacy_script_rules.py
index d7e085d4082f95c89ca990814bed7b6b9612cdea..5a71868f5d1c00c303ea915da21dc53cfac772f6 100644
--- a/test/functional/feature_legacy_script_rules.py
+++ b/test/functional/feature_legacy_script_rules.py
@@ -9,12 +9,7 @@ from test_framework.address import (
     SCRIPT_UNSPENDABLE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import (
-    VERSION_CHAIN_ID_BITS,
-    create_block,
-    create_coinbase,
-)
-from test_framework.hash import hash160
+from test_framework.blocktools import create_block, create_coinbase
 from test_framework.key import ECKey
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -39,11 +34,12 @@ from test_framework.script import (
     OP_SPLIT,
     OP_XOR,
     CScript,
+    SignatureHashForkId,
+    hash160,
 )
-from test_framework.signature_hash import SignatureHash, SignatureHashForkId
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
-from test_framework.util import assert_raises_rpc_error
+from test_framework.util import assert_equal, assert_raises_rpc_error
 
 
 class LegacyScriptRulesTest(BitcoinTestFramework):
@@ -72,7 +68,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
                 int(node.getbestblockhash(), 16),
                 create_coinbase(node.getblockcount() + 1),
             )
-            block.nVersion = VERSION_CHAIN_ID_BITS | 4
+            block.nVersion = 4
             block.vtx += [tx]
             block.hashMerkleRoot = block.calc_merkle_root()
             block.solve()
@@ -227,8 +223,8 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         spend_tx.vin = [CTxIn(COutPoint(int(txid, 16), checksig_out_idx))]
         pad_tx(spend_tx)
         script = CScript([OP_CHECKSIG])
-        (sighash, _) = SignatureHash(script, spend_tx, 0, 0x01)
-        txsig = private_key.sign_schnorr(sighash) + b"\x01"
+        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
+        txsig = private_key.sign_schnorr(sighash) + b"\x41"
         spend_tx.vin[0].scriptSig = CScript([txsig, public_key, script])
 
         # Regtest has SCRIPT_VERIFY_NULLFAIL, which comes in handy here
@@ -241,10 +237,6 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         peer = check_mined_tx_fails(
             node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
         )
-        # Schnorr sigs allowed on XEC with FORKID
-        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
-        txsig = private_key.sign_schnorr(sighash) + b"\x41"
-        spend_tx.vin[0].scriptSig = CScript([txsig, public_key, script])
         nonlegacy_node.sendrawtransaction(spend_tx.serialize().hex())
         check_mined_tx_success(nonlegacy_node, nonlegacy_peer, spend_tx)
 
@@ -253,8 +245,8 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         spend_tx.vin = [CTxIn(COutPoint(int(txid, 16), checksig_out_idx + 1))]
         pad_tx(spend_tx)
         script = CScript([OP_CHECKSIGVERIFY])
-        (sighash, _) = SignatureHash(script, spend_tx, 0, 0x01)
-        txsig = private_key.sign_schnorr(sighash) + b"\x01"
+        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
+        txsig = private_key.sign_schnorr(sighash) + b"\x41"
         spend_tx.vin[0].scriptSig = CScript([True, txsig, public_key, script])
 
         assert_raises_rpc_error(
@@ -266,9 +258,6 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         peer = check_mined_tx_fails(
             node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
         )
-        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
-        txsig = private_key.sign_schnorr(sighash) + b"\x41"
-        spend_tx.vin[0].scriptSig = CScript([True, txsig, public_key, script])
         nonlegacy_node.sendrawtransaction(spend_tx.serialize().hex())
         check_mined_tx_success(nonlegacy_node, nonlegacy_peer, spend_tx)
 
diff --git a/test/lint/check-doc.py b/test/lint/check-doc.py
index d3a3d27ff7ed1376826646a5294536608a320c35..1807dc0e96b5e704ab484a544e0bfde6ae88ef13 100755
--- a/test/lint/check-doc.py
+++ b/test/lint/check-doc.py
@@ -50,6 +50,7 @@ SET_FALSE_POSITIVE_UNDOCUMENTED = {
     "-chronikallowpause",
     "-chronikcors",
     "-dbcrashratio",
+    "-legacyscriptrules",
     "-enableminerfund",
     "-forcecompactdb",
     "-maxaddrtosend",
